pkg/util/validation/validate.go

```diff
@@ -336,7 +336,45 @@ func ValidateMetadata(validateMetrics *ValidateMetrics, cfg *Limits, userID stri
 	return nil
 }
 
+func validateNativeHistogramSchema(validateMetrics *ValidateMetrics, userID string, ls []cortexpb.LabelAdapter, histogramSample cortexpb.Histogram) (cortexpb.Histogram, error) {
+
+	// Exceed Min Exponential Schema limit.
+	if histogramSample.Schema < histogram.ExponentialSchemaMin {
+		validateMetrics.DiscardedSamples.WithLabelValues(nativeHistogramBucketCountLimitExceeded, userID).Inc()
+		return cortexpb.Histogram{}, newHistogramSchemaLimitExceededError(ls, int(histogram.ExponentialSchemaMin))
+	}
+
+	if histogramSample.IsFloatHistogram() {
+		// Exceed Max Exponential Schema limit for float histograms.
+		if histogramSample.Schema > histogram.ExponentialSchemaMax {
+			fh := cortexpb.FloatHistogramProtoToFloatHistogram(histogramSample)
+			fh = fh.ReduceResolution(histogram.ExponentialSchemaMax)
+			validateMetrics.HistogramSamplesReducedResolution.WithLabelValues(userID).Inc()
+			// After resolution reduced, convert the new float histogram to protobuf type again.
+			return cortexpb.FloatHistogramToHistogramProto(histogramSample.TimestampMs, fh), nil
```

Comment:
Maybe it is easier to just reject the sample if it has invalid schema. Bucket count is something configurable so maybe make sense to reduce resolution but I think it is fair to expect all samples to have valid schema.

Reply:
Prometheus seems to be doing this.
https://github.com/prometheus/prometheus/blob/24057883a14c597d70f46612e4b8fb5c9ad58d62/scrape/target.go#L415

Reply:
Thanks. Updated.